packages/sources/curve/src/endpoint/crypto.ts

```diff
@@ -0,0 +1,126 @@
+import registryExchangesABI from '../abis/registry_exchanges.json'
+import erc20ABI from '../abis/ERC20.json'
+import { Decimal } from 'decimal.js'
+
+export const supportedEndpoints = ['crypto']
+
+export const endpointResultPaths = {
+  crypto: 'rate',
+}
+
+export interface ResponseSchema {
+  pool: string
+  input: string
+  inputToken: string
+  inputDecimals: number
+  output: string
+  outputToken: string
+  outputDecimals: number
+  rate: number
+}
+
+export const inputParameters: InputParameters = {
+  from: ['base', 'from', 'coin'],
+  fromAddress: false,
+  fromDecimals: false,
+  to: ['quote', 'to', 'market'],
+  toAddress: false,
+  toDecimals: false,
+  amount: false,
+  resultPath: false,
+}
+
+export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
+  const validator = new Validator(request, inputParameters)
+  if (validator.error) throw validator.error
+
+  const jobRunID = validator.validated.id
+  const { address: from, decimals: fromDecimals } = await getTokenDetails(validator, 'from', config)
+  const { address: to, decimals: toDecimals } = await getTokenDetails(validator, 'to', config)
+  const inputAmount = validator.validated.data.amount || 1
+  const amount = BigNumber.from(inputAmount).mul(BigNumber.from(10).pow(fromDecimals))
+  const resultPath = validator.validated.data.resultPath
+
+  const [pool, output] = await getBestRate(from, to, amount, config)
+
+  const outputAmount = new Decimal(output.toString()).div(new Decimal(10).pow(toDecimals))
+  const rate = outputAmount.div(inputAmount)
+
+  const data: ResponseSchema = {
+    pool,
+    input: amount.toString(),
+    inputToken: from,
+    inputDecimals: fromDecimals,
+    output: output.toString(),
+    outputToken: to,
+    outputDecimals: toDecimals,
+    rate: rate.toNumber(),
+  }
+
+  const response = {
+    status: 200,
+    statusText: 'OK',
+    headers: {},
+    config: {},
+    data: data,
+  }
+  const result = Requester.validateResultNumber(response.data, [resultPath])
+
+  return Requester.success(jobRunID, Requester.withResult(response, result), config.verbose)
+}
+
+// getTokenDetails will find the address and number of decimal for a token.
+// `direction` is used to get the params in the request:
+//   - `{direction}` is the symbol of the token (to find pre-set token details)
+//   - `{direction}Address` is the token address as set in the request
+//   - `{direction}Decimals` is the number of decimals for the token as set in the request
+// The order of operations is as follows:
+//   address:
+//     1. Check if the address was provided in the request.
+//     2. If not, check the symbol in the request to see if we have pre-set the address for this symbol/network.
+//     3. If not, we assume the symbol param was actually the address.
+//   decimals:
+//     1. Check if the number of decimals was provided in the request.
+//     2. Query the contract at the address found above to see how many decimals it's set to.
```

Comment:
Tip:
If you're using vscode, you can start a TS doc line with `/**<ENTER>`, it'll automatically populate the comment with args and return types, and let you fill it in much quicker. Also lets future users using an editor with TS doc hinting support be able to view the comments as hints.

```suggestion
/**
 * getTokenDetails will find the address and number of decimal for a token.
 * 
 * The order of operations is as follows:
 *  - address:
 *     1. Check if the address was provided in the request.
 *     2. If not, check the symbol in the request to see if we have pre-set the address for this symbol/network.
 *     3. If not, we assume the symbol param was actually the address.
 *  - decimals:
 *     1. Check if the number of decimals was provided in the request.
 *     2. Query the contract at the address found above to see how many decimals it's set to.
 * @param validator The validation class to use
 * @param direction Used to get the params in the request
 * - `{direction}` is the symbol of the token (to find pre-set token details)
 * - `{direction}Address` is the token address as set in the request
 * - `{direction}Decimals` is the number of decimals for the token as set in the request
 * @param config Configuration to extract token decimals from
 */
```

Reply:
Cool :+1: